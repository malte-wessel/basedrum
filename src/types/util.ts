import { Observable } from 'rxjs';

export type UnfoldObservablesFromObject<T> = {
    [K in keyof T]: T[K] extends Observable<infer U> ? U : T[K];
};

export type UnfoldObservable<T> = T extends Observable<infer U>
    ? U
    : UnfoldObservablesFromObject<T>;

export type Dictionary<T> = {
    [key: string]: T;
};
