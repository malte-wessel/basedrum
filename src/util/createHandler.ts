import { Subject } from 'rxjs';

/**
 * Creates a handler function and a subject that emits when the handler is called
 */
export const createHandler = <E>() => {
    const subject = new Subject<E>();
    const handleEvent = (event: E) => subject.next(event);
    return [handleEvent, subject] as const;
};
