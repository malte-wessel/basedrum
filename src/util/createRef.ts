import { RefCallback } from 'preact';
import { BehaviorSubject } from 'rxjs';

/**
 * Creates a ref function and a BehaviourSubject that emits the referenced element
 */
export const createRef = <T extends HTMLElement>() => {
    const subject = new BehaviorSubject<T | null>(null);
    const ref: RefCallback<T> = el => {
        subject.next(el);
    };
    return [ref, subject] as const;
};
