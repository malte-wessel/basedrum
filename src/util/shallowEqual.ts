const hasOwn = Object.prototype.hasOwnProperty;

// Object.is polyfill
const is = (x: any, y: any) => {
    if (x === y) {
        return x !== 0 || y !== 0 || 1 / x === 1 / y;
    } else {
        // eslint-disable-next-line no-self-compare
        return x !== x && y !== y;
    }
};

export const shallowEqual = (a: any, b: any) => {
    if (is(a, b)) {
        return true;
    }

    if (
        typeof a !== 'object' ||
        a === null ||
        typeof b !== 'object' ||
        b === null
    ) {
        return false;
    }

    // eslint-disable-next-line no-self-compare
    if (Array.isArray(a) !== Array.isArray(b)) {
        return false;
    }

    const keysA = Object.keys(a);
    const keysB = Object.keys(b);

    if (keysA.length !== keysB.length) {
        return false;
    }

    for (let i = 0; i < keysA.length; i++) {
        if (!hasOwn.call(b, keysA[i]) || !is(a[keysA[i]], b[keysA[i]])) {
            return false;
        }
    }

    return true;
};
