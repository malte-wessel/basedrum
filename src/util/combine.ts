import { isObservable, Observable, combineLatest, of } from 'rxjs';
import { map } from 'rxjs/operators';
import { UnfoldObservable, Dictionary } from '../types/util';

type ObservableOrDictionary = Observable<any> | Dictionary<any>;

export function combine<T1>(t1: T1): Observable<UnfoldObservable<T1>>;
export function combine<T1, T2>(
    t1: T1,
    t2: T2,
): Observable<UnfoldObservable<T1> & UnfoldObservable<T2>>;
export function combine<T1, T2, T3>(
    t1: T1,
    t2: T2,
    t3: T3,
): Observable<
    UnfoldObservable<T1> & UnfoldObservable<T2> & UnfoldObservable<T3>
>;
export function combine<T1, T2, T3, T4>(
    t1: T1,
    t2: T2,
    t3: T3,
    t4: T4,
): Observable<
    UnfoldObservable<T1> &
        UnfoldObservable<T2> &
        UnfoldObservable<T3> &
        UnfoldObservable<T4>
>;
export function combine<T1, T2, T3, T4, T5>(
    t1: T1,
    t2: T2,
    t3: T3,
    t4: T4,
    t5: T5,
): Observable<
    UnfoldObservable<T1> &
        UnfoldObservable<T2> &
        UnfoldObservable<T3> &
        UnfoldObservable<T4> &
        UnfoldObservable<T5>
>;
export function combine<T1, T2, T3, T4, T5, T6>(
    t1: T1,
    t2: T2,
    t3: T3,
    t4: T4,
    t5: T5,
    t6: T6,
): Observable<
    UnfoldObservable<T1> &
        UnfoldObservable<T2> &
        UnfoldObservable<T3> &
        UnfoldObservable<T4> &
        UnfoldObservable<T5> &
        UnfoldObservable<T6>
>;
/**
 * `combine` merges data from Observables and plain objects that hold primitive values or Observables. The returned Observable emits whenever one of its input Observables emits.
 * @param streams
 */
export function combine(...streams: ObservableOrDictionary[]) {
    const sources = [];
    for (let i = 0, l = streams.length; i < l; i++) {
        const stream = streams[i];
        if (isObservable(stream)) {
            sources.push(stream);
        } else {
            for (const key in stream) {
                const value = stream[key];
                sources.push(
                    isObservable(value)
                        ? value.pipe(
                              map(value => ({
                                  [key]: value,
                              })),
                          )
                        : of({
                              [key]: value,
                          }),
                );
            }
        }
    }
    return combineLatest(...sources).pipe(
        map(values => Object.assign({}, ...values)),
    );
}
