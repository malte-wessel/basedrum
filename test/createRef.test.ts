import { createRef } from '../src/util/createRef';
describe('createRef', () => {
    it('should return a ref function and stream', () => {
        const el = document.createElement('div');
        const [ref, stream] = createRef<HTMLDivElement>();

        const emits: any = [];
        stream.subscribe(v => emits.push(v));

        ref(el);
        ref(null);
        expect(emits).toMatchSnapshot();
    });
});
