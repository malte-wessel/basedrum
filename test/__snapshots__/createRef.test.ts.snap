// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createRef should return a ref function and stream 1`] = `
Array [
  null,
  <div />,
  null,
]
`;
